src/study_groups.py

```
@@ -0,0 +1,712 @@
+            return helpers["err"]("You are already a member of this group", 400)
+
+        # Capacity check: atomically insert the membership row only when the
+        # group is below max_members to avoid races with concurrent accepts.
+        try:
+            await env.DB.prepare(
+                "INSERT INTO study_group_members (id,activity_id,user_id,role)"
+                " SELECT ?, ?, ?, 'member'"
+                " WHERE (SELECT COUNT(*) FROM study_group_members WHERE activity_id=?)"
+                "       < COALESCE((SELECT max_members FROM activities WHERE id=?), 1000000000)"
+            ).bind(helpers["new_id"](), invite.activity_id, user["id"], invite.activity_id, invite.activity_id).run()
+        except Exception as exc:
```

Ruff (0.15.20)

[warning] Do not catch blind exception: `Exception`

(BLE001)